sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchange.scala

```diff
@@ -103,16 +101,25 @@ case class ShuffleExchange(
     new ShuffledRowRDD(shuffleDependency, specifiedPartitionStartIndices)
   }
 
+  /**
+   * Caches the created ShuffleRowRDD so we can reuse that.
+   */
+  private var shuffleRDD: ShuffledRowRDD = null
+
   protected override def doExecute(): RDD[InternalRow] = attachTree(this, "execute") {
```

Comment:
attachTree is only used to generate better error message (show the plan), I think it do not matter here.